cirq-core/cirq/sim/act_on_args.py

```diff
@@ -70,11 +73,11 @@ def __init__(
             prng = cast(np.random.RandomState, np.random)
         if qubits is None:
             qubits = ()
-        if log_of_measurement_results is None:
-            log_of_measurement_results = {}
         self._set_qubits(qubits)
         self.prng = prng
-        self._log_of_measurement_results = log_of_measurement_results
+        # pylint: disable=line-too-long
+        self._classical_data = classical_data or value.ClassicalDataDictionaryStore(_measurements=log_of_measurement_results)  # type: ignore
```

Comment:
No, this is a good catch. I thought it was complaining about the `Optional`. I fixed this by making a copy of `log_of_measurement_results` with the required format. **HOWEVER** making a copy here strictly changes the behavior. Anyone dependent on `log_of_measurement_results` being passed by reference is now broken.

Note specifically that `stabilizer_ch_form_test.test_run` needed to be changed because of this. So if we think users might be doing the thing in that test, then we should worry about this. If not, then we shouldn't. (Since that test directly creates StabilizerStateChForm and ActOnStabilizerChFormArgs rather than just running CliffordSimulator, I'd assume this isn't a real-world use-case).

I could maybe mitigate this by temporarily adding a `old_log_of_measurement_results` argument/field to `ClassicalDataDictionaryStore` and duplicate the data there. Not sure it's worth it but I believe it would work.

Reply:
The pass-by-reference is required for separable state simulation. Each substate needs to refer to the same measurement data container. This PR changes that container from being the raw dictionary to being the ClassicalData object, and the simulators have been changed to share that object when states are separable (in create_partial_act_on_args).

The only real-world case in which this would break end users is if all of the following hold:
1. They have a custom simulator that extends `SimulatorBase`.
2. That simulator is capable of representing separable states via `ActOnArgsContainer` and `ActOnArgs.kron/factor/transpose`.
3. The circuits they're running on it perform measurements on one substate, and then do a classical control based on that measurement on a separate substate. I *think* this would end up breaking, but not even sure of that.

The fix for them is to update `create_partial_act_on_args` for their simulator as we did in this PR. (Which we decided was an allowed breaking change anyway).

TLDR: yeah I think we can mark it as breaking and move on.